File: Python/time/three_point_basis.py
from . import basis
from ..datastructures.tree import MetaRoot


class ContLinearScaling(basis.Scaling):
    """ Continuoues piecewise linear scaling function: `hat functions`.

    The function (l, n) corresponds to the hat function of node 2**-l * n.

    The fields `self.nbr_x` correspond to the neighbour on the left/right.
    The fields `self.child_x` correspond to the children on the left/mid/right.

    The field Element.phi_cont_lin object is ordered by labda, i.e. left, right.
    """
    __slots__ = [
        'nbr_left', 'nbr_right', 'child_left', 'child_mid', 'child_right'
    ]
    order = 1

    def __init__(self, labda, support, parents=None):
        super().__init__(labda, support=support, parents=parents)

        # TODO: We could read all these properties from the Element directly.
        self.nbr_left = None
        self.nbr_right = None

        self.child_left = None
        self.child_mid = None
        self.child_right = None

        l, n = labda
        if n > 0:
            assert support[0].phi_cont_lin[1] is None
            support[0].phi_cont_lin[1] = self
        if n < 2**l:
            assert support[-1].phi_cont_lin[0] is None
            support[-1].phi_cont_lin[0] = self

    def _update_children(self):
        """ Updates the children variable. Invoke after refining. """
        self.children = list(
            filter(None, [self.child_left, self.child_mid, self.child_right]))

    def refine(self):
        raise TypeError("Regular refinement of ContLinearScaling impossible!")

    def is_full(self):
        raise TypeError("ContLinearScaling functions cannot be `full' or not!")

    def refine_mid(self):
        if self.child_mid: return self.child_mid
        l, n = self.labda

        # Calculate the support of the refined hat.
        for elem in self.support:
            elem.refine()
        child_support = []
        if n > 0: child_support.append(self.support[0].children[1])
        if n < 2**l: child_support.append(self.support[-1].children[0])

        # Create element.
        child = ContLinearScaling((l + 1, 2 * n), child_support, [self])
        self.child_mid = child
        self._update_children()

        # Update nbrs.
        if self.child_left:
            self.child_left.nbr_right = child
            child.nbr_left = self.child_left

        if self.child_right:
            self.child_right.nbr_left = child
            child.nbr_right = self.child_right

        return child

    def refine_left(self):
        """ This refines the hat function on the left. """
        assert self.nbr_left
        if self.child_left:
            assert self.nbr_left.child_right == self.child_left
            return self.child_left

        # Some shortcuts for ease of reading.
        l, n = self.labda
        phi_left = self.nbr_left
        phi_right = self

        # Create child element.
        child = ContLinearScaling(
            (l + 1, n * 2 - 1),
            support=phi_right.support[0].children,
            parents=[phi_left, phi_right],
        )
        phi_left.child_right = child
        phi_right.child_left = child
        phi_left._update_children()
        phi_right._update_children()

        # Update nbrs.
        if phi_left.child_mid:
            phi_left.child_mid.nbr_right = child
            child.nbr_left = phi_left.child_mid

        if phi_right.child_mid:
            phi_right.child_mid.nbr_left = child
            child.nbr_right = phi_right.child_mid

        return child

    def refine_right(self):
        if self.child_right: return self.child_right
        return self.nbr_right.refine_left()

    def prolongate(self):
        l, n = self.labda
        result = [(self.refine_mid(), 1)]
        if n > 0: result.append((self.refine_left(), 0.5))
        if n < 2**l: result.append((self.refine_right(), 0.5))
        return result

    def restrict(self):
        if len(self.parents) == 1:
            return [(self.parents[0], 1)]
        else:
            return [(parent, 0.5) for parent in self.parents]

    @staticmethod
    def eval_mother(x, deriv=False):
        """ Evaluates the hat function on [-1,1] centered at 0. """
        left_mask = (-1 < x) & (x <= 0)
        right_mask = (0 < x) & (x < 1)

        if not deriv:
            return left_mask * (1 + x) + right_mask * (1 - x)
        else:
            return left_mask * 1 + right_mask * -1

    def eval(self, x, deriv=False):
        l, n = self.labda
        c = 2**l if deriv else 1.0
        return c * self.eval_mother(2**l * x - n, deriv)


class ThreePointWavelet(basis.Wavelet):
    """ Continuous piecewise linear basis, with three point wavelets.

    Scaling functions are simply the hat functions. A hat function on given
    level is indexed by the corresponding node index. A wavelet on level l >= 1
    is indexed by 0..2^l-1, corresponding to the odd nodes on level l.
    """
    __slots__ = []
    order = 1

    def __init__(self, labda, single_scale, parents=None):
        super().__init__(labda, single_scale=single_scale, parents=parents)

    def refine(self):
        if not self.children and self.level == 0:
            # Find all other wavelets on level 0 -- copy of scaling functions.
            parents = self.parents[0].children
            assert len(parents) == 2
            assert parents[0].labda == (0, 0)
            assert parents[1].labda == (0, 1)

            # Find the associated single_scale functions.
            mother_scalings = [
                parents[0].single_scale[0][0], parents[1].single_scale[0][0]
            ]

            # Create the mother wavelet.
            sq2 = 2**(1 / 2)
            child = ThreePointWavelet(
                (1, 0),
                single_scale=[(mother_scalings[0].refine_mid(), -sq2),
                              (mother_scalings[0].refine_right(), sq2),
                              (mother_scalings[1].refine_mid(), -sq2)],
                parents=parents)
            for parent in parents:
                parent.children = [child]
        elif not self.children and self.level > 0:
            phi_left, phi_mid, phi_right = [
                phi for phi, _ in self.single_scale
            ]
            l, n = self.labda
            scaling = 2**((l + 1) / 2)

            # First refine the left part
            phi_children = phi_left.refine_mid(), phi_mid.refine_left(
            ), phi_mid.refine_mid()
            if n == 0:
                single_scale = zip(
                    phi_children,
                    (-1 * scaling, 1 * scaling, -1 / 2 * scaling))
            else:
                single_scale = zip(
                    phi_children,
                    (-1 / 2 * scaling, 1 * scaling, -1 / 2 * scaling))

            child_left = ThreePointWavelet((l + 1, 2 * n), single_scale,
                                           [self])

            # Now refine the right part
            phi_children = phi_mid.refine_mid(), phi_mid.refine_right(
            ), phi_right.refine_mid()
            if n == 2**(l - 1) - 1:
                single_scale = zip(
                    phi_children,
                    (-1 / 2 * scaling, 1 * scaling, -1 * scaling))
            else:
                single_scale = zip(
                    phi_children,
                    (-1 / 2 * scaling, 1 * scaling, -1 / 2 * scaling))

            child_right = ThreePointWavelet((l + 1, 2 * n + 1), single_scale,
                                            [self])
            self.children = [child_left, child_right]
        return self.children

    def is_full(self):
        if self.level == 0: return len(self.children) == 1
        else: return len(self.children) == 2


class ThreePointBasis(basis.Basis):
    # Create mother scaling functions; roots of scaling tree.
    mother_scalings = [
        ContLinearScaling((0, 0), [basis.mother_element]),
        ContLinearScaling((0, 1), [basis.mother_element])
    ]
    mother_scalings[0].nbr_right = mother_scalings[1]
    mother_scalings[1].nbr_left = mother_scalings[0]

    # Create the root of the wavelet tree -- same as the mother scaling.
    roots_wavelet = [
        ThreePointWavelet((0, 0), single_scale=[(mother_scalings[0], 1)]),
        ThreePointWavelet((0, 1), single_scale=[(mother_scalings[1], 1)])
    ]

    # Create the metaroots
    metaroot_wavelet = MetaRoot(roots_wavelet)
    metaroot_scaling = MetaRoot(mother_scalings)

    def __init__(self):
        pass

File: C++/time/three_point_basis.cpp
#include "three_point_basis.hpp"

#include <cmath>
#include <iostream>
#include <memory>
namespace Time {

// Initialize static variables.
datastructures::Tree<ContLinearScalingFn> cont_lin_tree;
datastructures::Tree<ThreePointWaveletFn> three_point_tree;

// Metaroot constructor.
ContLinearScalingFn::ContLinearScalingFn() : ScalingFn<ContLinearScalingFn>() {
  auto scaling_left = make_child(
      /* parents */ std::vector{this},
      /* index */ 0,
      /* support */ std::vector{MotherElement()});
  auto scaling_right = make_child(
      /* parents */ std::vector{this},
      /* index */ 1,
      /* support */ std::vector{MotherElement()});

  scaling_left->nbr_right_ = scaling_right;
  scaling_right->nbr_left_ = scaling_left;
}

double ContLinearScalingFn::EvalMother(double t, bool deriv) const {
  int left_mask = (-1 < t) && (t <= 0);
  int right_mask = (0 < t) && (t < 1);

  if (deriv)
    return left_mask * 1 + right_mask * -1;
  else
    return left_mask * (1 + t) + right_mask * (1 - t);
}

ContLinearScalingFn *ContLinearScalingFn::RefineMiddle() {
  if (child_middle_) return child_middle_;
  auto [l, n] = labda();
  for (auto elem : support_) elem->Refine();

  std::vector<Element1D *> child_support;
  if (n > 0) child_support.push_back(support_[0]->children()[1]);
  if (n < (1 << l)) child_support.push_back(support_.back()->children()[0]);

  // Create child, and add accordingly.
  child_middle_ = make_child(
      /* parents */ std::vector{this},
      /* index */ 2 * n,
      /* support */ child_support);

  if (child_left_) {
    child_left_->nbr_right_ = child_middle_;
    child_middle_->nbr_left_ = child_left_;
  }
  if (child_right_) {
    child_right_->nbr_left_ = child_middle_;
    child_middle_->nbr_right_ = child_right_;
  }

  return child_middle_;
}

ContLinearScalingFn *ContLinearScalingFn::RefineLeft() {
  assert(nbr_left_);
  if (child_left_) return child_left_;
  support_[0]->Refine();
  auto [l, n] = labda();

  // Create child, and add accordingly.
  auto elems = support_[0]->children();
  child_left_ = make_child(
      /* parents */ std::vector{nbr_left_, this},
      /* index */ (2 * n - 1),
      /* support */ std::vector{elems[0], elems[1]});

  // Add this child to our left neighbour.
  nbr_left_->child_right_ = child_left_;
  nbr_left_->children_.push_back(child_left_);

  // Update neighbours of children.
  if (nbr_left_->child_middle_) {
    nbr_left_->child_middle_->nbr_right_ = child_left_;
    child_left_->nbr_left_ = nbr_left_->child_middle_;
  }
  if (child_middle_) {
    child_middle_->nbr_left_ = child_left_;
    child_left_->nbr_right_ = child_middle_;
  }

  return child_left_;
}

ThreePointWaveletFn::ThreePointWaveletFn() : WaveletFn<ThreePointWaveletFn>() {
  auto mother_scalings = cont_lin_tree.meta_root->children();
  assert(mother_scalings.size() == 2);
  for (size_t i = 0; i < 2; ++i) {
    make_child(
        /* parents */ std::vector{this},
        /* index */ i,
        /* single_scale */
        std::vector{std::pair{mother_scalings[i], 1.0}});
  }
}

bool ThreePointWaveletFn::Refine() {
  if (is_full()) return false;
  assert(children_.empty());

  if (level_ == 0) {
    // Find all wavelets on level 0 -- copy of scaling functions.
    auto parents = this->parents()[0]->children();
    assert(parents.size() == 2 && parents[0]->index_ == 0 &&
           parents[1]->index_ == 1);

    // Find the associated single scale functions
    std::vector<ContLinearScalingFn *> mother_scalings{
        parents[0]->single_scale_[0].first, parents[1]->single_scale_[0].first};

    // Refine the mother scaling function(s).
    mother_scalings[0]->RefineMiddle();
    mother_scalings[0]->RefineRight();
    mother_scalings[1]->RefineMiddle();

    assert(mother_scalings[0]->child_middle_);
    assert(mother_scalings[0]->child_right_);
    assert(mother_scalings[1]->child_middle_);
    assert(parents[0]);
    assert(parents[1]);

    // Create the mother wavelet.
    double sq2 = sqrt(2);
    auto child = parents[0]->make_child(
        /* parents */ std::vector{parents[0], parents[1]},
        /* index */ 0,
        /* single_scale */
        std::vector{std::pair{mother_scalings[0]->child_middle_, -sq2},
                    std::pair{mother_scalings[0]->child_right_, sq2},
                    std::pair{mother_scalings[1]->child_middle_, -sq2}});
    parents[1]->children_ = {child};
  } else {
    assert(level_ > 0);
    auto phi_left = single_scale_[0].first;
    auto phi_middle = single_scale_[1].first;
    auto phi_right = single_scale_[2].first;

    // Refine the necessary scaling functions.
    phi_left->RefineMiddle();
    phi_middle->RefineLeft();
    phi_middle->RefineMiddle();
    phi_middle->RefineRight();
    phi_right->RefineMiddle();

    // First refine the left part.
    auto [l, n] = labda();
    double scaling = pow(2, (l + 1) / 2.0);
    std::vector<ContLinearScalingFn *> phi_children{phi_left->child_middle_,
                                                    phi_middle->child_left_,
                                                    phi_middle->child_middle_};

    if (n == 0) {
      make_child(
          /* parents */ std::vector{this},
          /* index */ 2 * index_,
          /* single_scale */
          std::vector{std::pair{phi_children[0], -scaling},
                      std::pair{phi_children[1], scaling},
                      std::pair{phi_children[2], -0.5 * scaling}});
    } else {
      make_child(
          /* parents */ std::vector{this},
          /* index */ 2 * index_,
          /* single_scale */
          std::vector{std::pair{phi_children[0], -0.5 * scaling},
                      std::pair{phi_children[1], scaling},
                      std::pair{phi_children[2], -0.5 * scaling}});
    }

    // Now refine ther right part
    phi_children = {phi_middle->child_middle_, phi_middle->child_right_,
                    phi_right->child_middle_};

    if (n == (1 << (l - 1)) - 1) {
      make_child(
          /* parents */ std::vector{this},
          /* index */ 2 * index_ + 1,
          /* single_scale */
          std::vector{std::pair{phi_children[0], -0.5 * scaling},
                      std::pair{phi_children[1], scaling},
                      std::pair{phi_children[2], -scaling}});
    } else {
      make_child(
          /* parents */ std::vector{this},
          /* index */ 2 * index_ + 1,
          /* single_scale */
          std::vector{std::pair{phi_children[0], -0.5 * scaling},
                      std::pair{phi_children[1], scaling},
                      std::pair{phi_children[2], -0.5 * scaling}});
    }
  }
  return true;
}

bool ThreePointWaveletFn::is_full() const {
  if (level_ == 0)
    return children_.size() == 1;
  else
    return children_.size() == 2;
}

}  // namespace Time
